### PR TITLE
HPA: calculate sidecar container resource in pod autoscaler

### LINKED ISSUE #119991: HPA: Calculate sidecar container resource in pod autoscaler
### What would you like to be added?

Calculate sidecar containers resource in pod autoscaler.
https://github.com/kubernetes/kubernetes/blob/4f874a224a31a5dc5745a0c140d3769bb178f811/pkg/controller/podautoscaler/replica_calculator.go#L424

### Why is this needed?

From 1.28 user could run **Sidecar containers** in `initContainers`, those sidecar containers remain running during the entire life of the Pod, it should be calculated in [pod autoscaler](https://github.com/kubernetes/kubernetes/blob/4f874a224a31a5dc5745a0c140d3769bb178f811/pkg/controller/podautoscaler/replica_calculator.go#L424).

If missing the sidecar resources, we will get a wrong pod resource utilization.

We should also correct the related Cadvisor metrics.

### PR DESCRIPTION
**Pull request by qingwave**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
From 1.28 user could run Sidecar containers in initContainers, those sidecar containers remain running during the entire life of the Pod, it should be calculated in [pod autoscaler](https://github.com/kubernetes/kubernetes/blob/4f874a224a31a5dc5745a0c140d3769bb178f811/pkg/controller/podautoscaler/replica_calculator.go#L424).

If missing the sidecar resources, we will get a wrong pod resource utilization.

#### Which issue(s) this PR fixes:

Fixes #119991

#### Does this PR introduce a user-facing change?

```release-note
Calculate restartable init containers resource in pod autoscaler
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Review by gjkim42 — COMMENTED** [review]
Thanks for catching this.
/lgtm

File: pkg/controller/podautoscaler/replica_calculator_test.go [hunk 1]
```diff
@@ -1999,3 +1999,112 @@ func TestGroupPods(t *testing.T) {
...
+					},
+				},
+			}},
+			container:        "",
+			resource:         v1.ResourceCPU,
+			expectedRequests: nil,
+			expectedError:    fmt.Errorf("missing request for %s in container %s of Pod %s", v1.ResourceCPU, "container1", testPod),
+		},
+		{
+			name: "pod with sidecar containers",
```
Comment (gjkim42) [inline pkg/controller/podautoscaler/replica_calculator_test.go:2080]: We decided to call an init container with always restartPolicy as a restartable init container.

Comment (qingwave) [reply pkg/controller/podautoscaler/replica_calculator_test.go]: done

**Review by gjkim42 — COMMENTED** [review]
/lgtm

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm

**Pull request merged**